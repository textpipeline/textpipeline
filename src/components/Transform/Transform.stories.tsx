import { storiesOf } from '@storybook/react';
import React from 'react';

import { testTransform } from '../../fixtures';
import routes from '../../routes';
import Transform from './Transform';

storiesOf('Transform', module).add('default', () => (
  <Transform
    name={testTransform.name}
    ossHref={`${routes.oss.path}/${testTransform.slug}`}
    inputType={testTransform.inputType}
    outputType={testTransform.outputType}
    defaultOutput={testTransform.defaultOutput}
    execute={testTransform.execute}
  />
));
