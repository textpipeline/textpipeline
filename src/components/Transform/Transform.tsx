import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import CircularProgress from '@material-ui/core/CircularProgress';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import AssignmentIcon from '@material-ui/icons/Assignment';
import RepeatIcon from '@material-ui/icons/Repeat';
import copy from 'clipboard-copy';
import React, { ChangeEvent, useState } from 'react';

import { Transform as TextTransform } from '../../transforms/types';
import Heading from '../Heading';
import InternalLink from '../InternalLink';

const textFieldRowCount = 12;

export interface TransformProps {
  readonly name: TextTransform['name'];
  readonly ossHref: string;
  readonly inputType: TextTransform['inputType'];
  readonly outputType: TextTransform['outputType'];
  readonly execute: TextTransform['execute'];
}

const Transform: React.FC<TransformProps> = ({ name, ossHref, inputType, outputType, execute }) => {
  const [inputValue, setInputValue] = useState('');
  const [outputValue, setOutputValue] = useState('');
  const [error, setError] = useState<String | string | null>(null);
  const [copied, setCopied] = useState(false);
  const [loading, setLoading] = useState(false);

  const isValid = !error;
  const inputHelperText = isValid
    ? "Once you've finished click the button to transform"
    : `There was an error! ${error}`;

  const onChange = (event: ChangeEvent<HTMLInputElement>) => {
    const newInputValue = event.target.value;
    setInputValue(newInputValue || '');
  };

  const onTransformClick = async () => {
    if (inputValue === '') {
      setCopied(false);
      setOutputValue('');
      setError(null);
      return;
    }
    setCopied(false);
    setLoading(true);
    try {
      const output = await execute(inputValue);
      setOutputValue(output);
    } catch (err) {
      if (err instanceof Error) {
        setError(err.message);
      } else if (typeof err === 'string' || err instanceof String) {
        setError(err);
      } else {
        setError('Unable to transform text.');
      }
    } finally {
      setLoading(false);
    }
  };

  const onClearClick = () => {
    setInputValue('');
    setOutputValue('');
    setError(null);
    setCopied(false);
  };

  const onCopyToClipboardClick = () => {
    copy(outputValue);
    setCopied(true);
  };

  return (
    <>
      <Box display="flex" flexWrap="nowrap" justifyContent="space-between" alignItems="center">
        <Heading level={2}>{name}</Heading>
        <InternalLink href={ossHref}>about this transform</InternalLink>
      </Box>

      <Box my={3}>
        <form noValidate autoComplete="off">
          <TextField
            id="input-box"
            label="Text to transform"
            placeholder={`Enter ${inputType} here`}
            helperText={inputHelperText}
            error={!isValid}
            variant="outlined"
            rows={textFieldRowCount}
            value={inputValue}
            onChange={onChange}
            color="primary"
            autoFocus
            fullWidth
            multiline
            required
            disabled={loading}
          />
          <Box display="flex" justifyContent="flex-end" my={3}>
            {loading && <CircularProgress />}
            <Box mx={1}>
              <Button
                id="transform-button"
                onClick={onTransformClick}
                variant="contained"
                color="primary"
                size="large"
                disabled={!inputValue || loading}
                endIcon={<RepeatIcon />}
                disableElevation
              >
                Transform
              </Button>
            </Box>
            <Button
              id="clear-button"
              onClick={onClearClick}
              variant="outlined"
              disableElevation
              size="large"
              disabled={!inputValue || loading}
            >
              Clear
            </Button>
          </Box>
          <TextField
            id="output-box"
            label="Transformed text"
            placeholder={`${outputType} will appear here`}
            variant="outlined"
            rows={textFieldRowCount}
            value={outputValue}
            color="primary"
            InputProps={{
              readOnly: true,
            }}
            fullWidth
            multiline
            disabled={loading}
          />
          <Box display="flex" justifyContent="flex-end" my={3}>
            {copied && (
              <Box mx={1} color="success.main">
                <Typography id="copied-indicator">Copied!</Typography>
              </Box>
            )}
            <Button
              id="copy-button"
              onClick={onCopyToClipboardClick}
              variant="contained"
              color="primary"
              size="small"
              disabled={!outputValue}
              endIcon={<AssignmentIcon />}
              disableElevation
            >
              Copy
            </Button>
          </Box>
        </form>
      </Box>
    </>
  );
};

export default Transform;
