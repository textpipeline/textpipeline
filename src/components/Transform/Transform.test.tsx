import CircularProgress from '@material-ui/core/CircularProgress';
import copy from 'clipboard-copy';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';

import Transform, { TransformProps } from './Transform';

jest.mock('clipboard-copy');

class Deferred {
  promise: Promise<void>;
  reject: (error: any) => void = () => {};
  resolve: () => void = () => {};

  constructor() {
    this.promise = new Promise((resolve, reject) => {
      this.reject = reject;
      this.resolve = resolve;
    });
  }
}

const getInputBox = (rendered: ShallowWrapper<TransformProps>) => rendered.find('#input-box');
const getOutputBox = (rendered: ShallowWrapper<TransformProps>) => rendered.find('#output-box');
const getTransformButton = (rendered: ShallowWrapper<TransformProps>) => rendered.find('#transform-button');
const getClearButton = (rendered: ShallowWrapper<TransformProps>) => rendered.find('#clear-button');
const getCopyButton = (rendered: ShallowWrapper<TransformProps>) => rendered.find('#copy-button');
const getCopiedIndicator = (rendered: ShallowWrapper<TransformProps>) => rendered.find('#copied-indicator');
const getLoadingIndicator = (rendered: ShallowWrapper<TransformProps>) => rendered.find(CircularProgress);

describe('Transform', () => {
  let execute: TransformProps['execute'];
  let rendered: ShallowWrapper<TransformProps>;

  beforeEach(() => {
    execute = jest.fn((input: string) => Promise.resolve(`${input} - transformed to a Bar!`));
    rendered = shallow(<Transform name="Foo to Bar" ossHref="#" inputType="FOO" outputType="BAR" execute={execute} />);
  });

  it('renders correctly', () => {
    expect(rendered).toMatchSnapshot();
  });

  describe('before the user has entered text in the input box', () => {
    beforeEach(() => {
      getInputBox(rendered).simulate('change', { target: { value: '' } });
    });

    it('renders the transform button disabled', () => {
      expect(getTransformButton(rendered).prop('disabled')).toBe(true);
    });

    it('renders the clear button disabled', () => {
      expect(getClearButton(rendered).prop('disabled')).toBe(true);
    });

    it('renders the copy-to-clipboard button disabled', () => {
      expect(getCopyButton(rendered).prop('disabled')).toBe(true);
    });
  });

  describe('when text is typed into the input box', () => {
    beforeEach(() => {
      getInputBox(rendered).simulate('change', { target: { value: 'stuff' } });
    });

    it('updates the state with the text', () => {
      expect(getInputBox(rendered).prop('value')).toEqual('stuff');
    });
  });

  describe('when no text was input and the transform button is clicked', () => {
    beforeEach(() => {
      getInputBox(rendered).simulate('change', { target: { value: '' } });
      getTransformButton(rendered).simulate('click');
    });

    it('does not place any text in the output box', () => {
      expect(getOutputBox(rendered).prop('value')).toEqual('');
    });

    it('resets the copied state', () => {
      expect(getCopiedIndicator(rendered).length).toEqual(0);
    });
  });

  describe('after the user has entered text in the input box', () => {
    beforeEach(() => {
      getInputBox(rendered).simulate('change', { target: { value: 'stuff' } });
    });

    describe('when the clear button is clicked', () => {
      beforeEach(() => {
        getClearButton(rendered).simulate('click');
      });

      it('resets the input box to empty', () => {
        expect(getInputBox(rendered).prop('value')).toEqual('');
      });

      it('renders the transform button disabled', () => {
        expect(getTransformButton(rendered).prop('disabled')).toBe(true);
      });

      it('renders the clear button disabled', () => {
        expect(getClearButton(rendered).prop('disabled')).toBe(true);
      });

      it('renders the copy-to-clipboard button disabled', () => {
        expect(getCopyButton(rendered).prop('disabled')).toBe(true);
      });
    });

    describe('when the transform will fail', () => {
      let deferred: Deferred;

      beforeEach(() => {
        deferred = new Deferred();
        (execute as jest.Mock).mockReturnValue(deferred.promise);
      });

      describe('when the transform button is clicked', () => {
        beforeEach(() => {
          getTransformButton(rendered).simulate('click');
        });

        describe('before the transform completes', () => {
          it('displays a loading indicator', () => {
            expect(getLoadingIndicator(rendered).length).toEqual(1);
          });
        });

        describe.each([
          { error: 'boom', expectedMessage: 'There was an error! boom' },
          { error: new Error('boom'), expectedMessage: 'There was an error! boom' },
          { error: 42, expectedMessage: 'There was an error! Unable to transform text.' },
        ])('after the transform fails', ({ error, expectedMessage }) => {
          beforeEach(() => {
            deferred.reject(error);
          });

          it('no longer displays a loading indicator', async () => {
            try {
              await deferred.promise;
            } catch {
              expect(getLoadingIndicator(rendered).length).toEqual(0);
            }
          });

          it('sets the error state', () => {
            expect(getInputBox(rendered).prop('helperText')).toEqual(expectedMessage);
          });

          it('renders correctly', () => {
            expect(rendered).toMatchSnapshot();
          });

          describe('when the clear button is clicked', () => {
            beforeEach(() => {
              getClearButton(rendered).simulate('click');
            });

            it('resets the input box to empty', () => {
              expect(getInputBox(rendered).prop('value')).toEqual('');
            });

            it('resets the output box to empty', () => {
              expect(getOutputBox(rendered).prop('value')).toEqual('');
            });

            it('renders the transform button disabled', () => {
              expect(getTransformButton(rendered).prop('disabled')).toBe(true);
            });

            it('renders the clear button disabled', () => {
              expect(getClearButton(rendered).prop('disabled')).toBe(true);
            });

            it('renders the copy-to-clipboard button disabled', () => {
              expect(getCopyButton(rendered).prop('disabled')).toBe(true);
            });

            it('clears the error state', () => {
              expect(getInputBox(rendered).prop('helperText')).toEqual(
                "Once you've finished click the button to transform"
              );
            });
          });
        });
      });
    });

    describe('when the transform will succeed', () => {
      let deferred: Deferred;

      beforeEach(() => {
        deferred = new Deferred();
        (execute as jest.Mock).mockReturnValue(deferred.promise);
      });

      describe('when the transform button is clicked', () => {
        beforeEach(() => {
          getTransformButton(rendered).simulate('click');
        });

        describe('before the transform completes', () => {
          it('displays a loading indicator', () => {
            expect(getLoadingIndicator(rendered).length).toEqual(1);
          });
        });

        describe('after the transform succeeds', () => {
          beforeEach(() => {
            deferred.resolve();
          });

          it('no longer displays a loading indicator', async () => {
            await deferred.promise;
            expect(getLoadingIndicator(rendered).length).toEqual(0);
          });

          describe('when the copy-to-clipboard button is clicked', () => {
            beforeEach(() => {
              getCopyButton(rendered).simulate('click');
            });

            it('copies the text to the clipboard', () => {
              expect(copy).toHaveBeenCalled();
            });

            it('displays an indication that the text was copied', () => {
              expect(getCopiedIndicator(rendered).length).toEqual(1);
            });

            describe('when the transform button is clicked again', () => {
              beforeEach(() => {
                getTransformButton(rendered).simulate('click');
              });

              it('clears the copied state', () => {
                expect(getCopiedIndicator(rendered).length).toEqual(0);
              });
            });
          });

          describe('when the clear button is clicked', () => {
            beforeEach(() => {
              getClearButton(rendered).simulate('click');
            });

            it('resets the input box to empty', () => {
              expect(getInputBox(rendered).prop('value')).toEqual('');
            });

            it('resets the output box to empty', () => {
              expect(getOutputBox(rendered).prop('value')).toEqual('');
            });

            it('renders the transform button disabled', () => {
              expect(getTransformButton(rendered).prop('disabled')).toBe(true);
            });

            it('renders the clear button disabled', () => {
              expect(getClearButton(rendered).prop('disabled')).toBe(true);
            });

            it('renders the copy-to-clipboard button disabled', () => {
              expect(getCopyButton(rendered).prop('disabled')).toBe(true);
            });

            it('clears the error state', () => {
              expect(getInputBox(rendered).prop('helperText')).toEqual(
                "Once you've finished click the button to transform"
              );
            });
          });
        });
      });
    });
  });
});
