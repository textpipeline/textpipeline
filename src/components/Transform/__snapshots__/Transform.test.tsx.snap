// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transform after the user has entered text in the input box when the transform will fail when the transform button is clicked after the transform fails renders correctly 1`] = `
<Fragment>
  <Styled(MuiBox)
    alignItems="center"
    display="flex"
    flexWrap="nowrap"
    justifyContent="space-between"
  >
    <Heading
      level={2}
    >
      Foo to Bar
    </Heading>
    <ForwardRef
      href="#"
    >
      about this transform
    </ForwardRef>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    my={3}
  >
    <form
      autoComplete="off"
      noValidate={true}
    >
      <WithStyles(ForwardRef(TextField))
        autoFocus={true}
        color="primary"
        disabled={false}
        error={true}
        fullWidth={true}
        helperText="There was an error! boom"
        id="input-box"
        label="Text to transform"
        multiline={true}
        onChange={[Function]}
        placeholder="Enter FOO here"
        required={true}
        rows={12}
        value="stuff"
        variant="outlined"
      />
      <Styled(MuiBox)
        display="flex"
        justifyContent="flex-end"
        my={3}
      >
        <Styled(MuiBox)
          mx={1}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            disableElevation={true}
            disabled={false}
            endIcon={<UNDEFINED />}
            id="transform-button"
            onClick={[Function]}
            size="large"
            variant="contained"
          >
            Transform
          </WithStyles(ForwardRef(Button))>
        </Styled(MuiBox)>
        <WithStyles(ForwardRef(Button))
          disableElevation={true}
          disabled={false}
          id="clear-button"
          onClick={[Function]}
          size="large"
          variant="outlined"
        >
          Clear
        </WithStyles(ForwardRef(Button))>
      </Styled(MuiBox)>
      <WithStyles(ForwardRef(TextField))
        InputProps={
          Object {
            "readOnly": true,
          }
        }
        color="primary"
        disabled={false}
        fullWidth={true}
        id="output-box"
        label="Transformed text"
        multiline={true}
        placeholder="BAR will appear here"
        rows={12}
        value=""
        variant="outlined"
      />
      <Styled(MuiBox)
        display="flex"
        justifyContent="flex-end"
        my={3}
      >
        <WithStyles(ForwardRef(Button))
          color="primary"
          disableElevation={true}
          disabled={true}
          endIcon={<UNDEFINED />}
          id="copy-button"
          onClick={[Function]}
          size="small"
          variant="contained"
        >
          Copy
        </WithStyles(ForwardRef(Button))>
      </Styled(MuiBox)>
    </form>
  </Styled(MuiBox)>
</Fragment>
`;

exports[`Transform after the user has entered text in the input box when the transform will fail when the transform button is clicked after the transform fails renders correctly 2`] = `
<Fragment>
  <Styled(MuiBox)
    alignItems="center"
    display="flex"
    flexWrap="nowrap"
    justifyContent="space-between"
  >
    <Heading
      level={2}
    >
      Foo to Bar
    </Heading>
    <ForwardRef
      href="#"
    >
      about this transform
    </ForwardRef>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    my={3}
  >
    <form
      autoComplete="off"
      noValidate={true}
    >
      <WithStyles(ForwardRef(TextField))
        autoFocus={true}
        color="primary"
        disabled={false}
        error={true}
        fullWidth={true}
        helperText="There was an error! boom"
        id="input-box"
        label="Text to transform"
        multiline={true}
        onChange={[Function]}
        placeholder="Enter FOO here"
        required={true}
        rows={12}
        value="stuff"
        variant="outlined"
      />
      <Styled(MuiBox)
        display="flex"
        justifyContent="flex-end"
        my={3}
      >
        <Styled(MuiBox)
          mx={1}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            disableElevation={true}
            disabled={false}
            endIcon={<UNDEFINED />}
            id="transform-button"
            onClick={[Function]}
            size="large"
            variant="contained"
          >
            Transform
          </WithStyles(ForwardRef(Button))>
        </Styled(MuiBox)>
        <WithStyles(ForwardRef(Button))
          disableElevation={true}
          disabled={false}
          id="clear-button"
          onClick={[Function]}
          size="large"
          variant="outlined"
        >
          Clear
        </WithStyles(ForwardRef(Button))>
      </Styled(MuiBox)>
      <WithStyles(ForwardRef(TextField))
        InputProps={
          Object {
            "readOnly": true,
          }
        }
        color="primary"
        disabled={false}
        fullWidth={true}
        id="output-box"
        label="Transformed text"
        multiline={true}
        placeholder="BAR will appear here"
        rows={12}
        value=""
        variant="outlined"
      />
      <Styled(MuiBox)
        display="flex"
        justifyContent="flex-end"
        my={3}
      >
        <WithStyles(ForwardRef(Button))
          color="primary"
          disableElevation={true}
          disabled={true}
          endIcon={<UNDEFINED />}
          id="copy-button"
          onClick={[Function]}
          size="small"
          variant="contained"
        >
          Copy
        </WithStyles(ForwardRef(Button))>
      </Styled(MuiBox)>
    </form>
  </Styled(MuiBox)>
</Fragment>
`;

exports[`Transform after the user has entered text in the input box when the transform will fail when the transform button is clicked after the transform fails renders correctly 3`] = `
<Fragment>
  <Styled(MuiBox)
    alignItems="center"
    display="flex"
    flexWrap="nowrap"
    justifyContent="space-between"
  >
    <Heading
      level={2}
    >
      Foo to Bar
    </Heading>
    <ForwardRef
      href="#"
    >
      about this transform
    </ForwardRef>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    my={3}
  >
    <form
      autoComplete="off"
      noValidate={true}
    >
      <WithStyles(ForwardRef(TextField))
        autoFocus={true}
        color="primary"
        disabled={false}
        error={true}
        fullWidth={true}
        helperText="There was an error! Unable to transform text."
        id="input-box"
        label="Text to transform"
        multiline={true}
        onChange={[Function]}
        placeholder="Enter FOO here"
        required={true}
        rows={12}
        value="stuff"
        variant="outlined"
      />
      <Styled(MuiBox)
        display="flex"
        justifyContent="flex-end"
        my={3}
      >
        <Styled(MuiBox)
          mx={1}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            disableElevation={true}
            disabled={false}
            endIcon={<UNDEFINED />}
            id="transform-button"
            onClick={[Function]}
            size="large"
            variant="contained"
          >
            Transform
          </WithStyles(ForwardRef(Button))>
        </Styled(MuiBox)>
        <WithStyles(ForwardRef(Button))
          disableElevation={true}
          disabled={false}
          id="clear-button"
          onClick={[Function]}
          size="large"
          variant="outlined"
        >
          Clear
        </WithStyles(ForwardRef(Button))>
      </Styled(MuiBox)>
      <WithStyles(ForwardRef(TextField))
        InputProps={
          Object {
            "readOnly": true,
          }
        }
        color="primary"
        disabled={false}
        fullWidth={true}
        id="output-box"
        label="Transformed text"
        multiline={true}
        placeholder="BAR will appear here"
        rows={12}
        value=""
        variant="outlined"
      />
      <Styled(MuiBox)
        display="flex"
        justifyContent="flex-end"
        my={3}
      >
        <WithStyles(ForwardRef(Button))
          color="primary"
          disableElevation={true}
          disabled={true}
          endIcon={<UNDEFINED />}
          id="copy-button"
          onClick={[Function]}
          size="small"
          variant="contained"
        >
          Copy
        </WithStyles(ForwardRef(Button))>
      </Styled(MuiBox)>
    </form>
  </Styled(MuiBox)>
</Fragment>
`;

exports[`Transform renders correctly 1`] = `
<Fragment>
  <Styled(MuiBox)
    alignItems="center"
    display="flex"
    flexWrap="nowrap"
    justifyContent="space-between"
  >
    <Heading
      level={2}
    >
      Foo to Bar
    </Heading>
    <ForwardRef
      href="#"
    >
      about this transform
    </ForwardRef>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    my={3}
  >
    <form
      autoComplete="off"
      noValidate={true}
    >
      <WithStyles(ForwardRef(TextField))
        autoFocus={true}
        color="primary"
        disabled={false}
        error={false}
        fullWidth={true}
        helperText="Once you've finished click the button to transform"
        id="input-box"
        label="Text to transform"
        multiline={true}
        onChange={[Function]}
        placeholder="Enter FOO here"
        required={true}
        rows={12}
        value=""
        variant="outlined"
      />
      <Styled(MuiBox)
        display="flex"
        justifyContent="flex-end"
        my={3}
      >
        <Styled(MuiBox)
          mx={1}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            disableElevation={true}
            disabled={true}
            endIcon={<UNDEFINED />}
            id="transform-button"
            onClick={[Function]}
            size="large"
            variant="contained"
          >
            Transform
          </WithStyles(ForwardRef(Button))>
        </Styled(MuiBox)>
        <WithStyles(ForwardRef(Button))
          disableElevation={true}
          disabled={true}
          id="clear-button"
          onClick={[Function]}
          size="large"
          variant="outlined"
        >
          Clear
        </WithStyles(ForwardRef(Button))>
      </Styled(MuiBox)>
      <WithStyles(ForwardRef(TextField))
        InputProps={
          Object {
            "readOnly": true,
          }
        }
        color="primary"
        disabled={false}
        fullWidth={true}
        id="output-box"
        label="Transformed text"
        multiline={true}
        placeholder="BAR will appear here"
        rows={12}
        value=""
        variant="outlined"
      />
      <Styled(MuiBox)
        display="flex"
        justifyContent="flex-end"
        my={3}
      >
        <WithStyles(ForwardRef(Button))
          color="primary"
          disableElevation={true}
          disabled={true}
          endIcon={<UNDEFINED />}
          id="copy-button"
          onClick={[Function]}
          size="small"
          variant="contained"
        >
          Copy
        </WithStyles(ForwardRef(Button))>
      </Styled(MuiBox)>
    </form>
  </Styled(MuiBox)>
</Fragment>
`;
